WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModel.swift

```diff
@@ -19,30 +19,185 @@ final class ReviewsDashboardCardViewModel: ObservableObject {
+        productsResultsController.predicate = predicates
+    }
+}
+
+// MARK: - Private helpers
+private extension ReviewsDashboardCardViewModel {
+    @MainActor
+    func loadReviews() async throws -> [ProductReview] {
+        try await withCheckedThrowingContinuation { continuation in
+            stores.dispatch(ProductReviewAction.synchronizeProductReviews(siteID: siteID,
+                                                                          pageNumber: 1,
+                                                                          pageSize: Constants.numberOfItems) { result in
+                continuation.resume(with: result)
+            })
+        }
+    }
+
+    /// Updates data
+    @MainActor
+    func updateReviewsResults() async {
+        let reviews = productReviewsResultsController.fetchedObjects.prefix(Constants.numberOfItems)
+        let productIDs = reviews.map { $0.productID }
+
+        if productIDs.isNotEmpty {
+            updateProductsResultsControllerPredicate(with: productIDs)
+            do {
+                try await fetchProducts(for: productIDs)
+            } catch {
+                ServiceLocator.crashLogging.logError(error)
+            }
+        }
+
+        data = reviews
+            .map { review in
+                let product = reviewProducts.first { $0.productID == review.productID }
+                // TODO: also fetch notification
+                return ReviewViewModel(review: review, product: product, notification: nil)
+            }
+    }
+
+    /// Get products from storage if available, if not then fetch remotely.
+    ///
+    @MainActor
+    private func fetchProducts(for productIDs: [Int64]) async throws {
+        try productsResultsController.performFetch()
+
+        // Check if all productIDs are available in storage
+        let allProductsAvailable = productIDs.allSatisfy { productID in
+            reviewProducts.contains { $0.productID == productID }
+        }
+
+        if !allProductsAvailable {
+            await loadReviewProducts(for: productIDs)
+        }
+    }
+
+    @MainActor
+    func loadReviewProducts(for reviewProductIDs: [Int64]) async {
+        syncingData = true
+        syncingError = nil
+
+        do {
+            // Ignoring the result from remote as we're using storage as the single source of truth
+            _ = try await retrieveProducts(for: reviewProductIDs)
+        } catch {
+            syncingError = error
+        }
+        syncingData = false
+    }
+
+    @MainActor
+    func retrieveProducts(for reviewProductIDs: [Int64]) async throws -> (products: [Product], hasNextPage: Bool) {
+        try await withCheckedThrowingContinuation { continuation in
+            stores.dispatch(ProductAction.retrieveProducts(siteID: siteID,
+                                                           productIDs: reviewProductIDs
+                                                          ) { result in
+                continuation.resume(with: result)
```

Comment:
Non blocking suggestion:
We could replace this with the following to skip returning stuff from this function.

```swift
switch result {
 case .success:
  continuation.resume()
 case .failure(let error):
  continuation.resume(throwing: error)
}
```

The same applies to `func loadReviews() async throws -> [ProductReview]` method as well.

Reply:
Thanks, I added the updates for these functions on https://github.com/woocommerce/woocommerce-ios/pull/12752/commits/40bb44abeae6843fcef40bbe860838ddd7a93dc0